src/java.base/share/classes/java/util/AbstractMap.java

```diff
@@ -348,46 +318,41 @@ public void clear() {
      * method will not all return the same set.
      */
     public Set<K> keySet() {
-        Set<K> ks = keySet;
-        if (ks == null) {
-            ks = new AbstractSet<K>() {
-                public Iterator<K> iterator() {
-                    return new Iterator<K>() {
-                        private Iterator<Entry<K,V>> i = entrySet().iterator();
-
-                        public boolean hasNext() {
-                            return i.hasNext();
-                        }
-
-                        public K next() {
-                            return i.next().getKey();
-                        }
-
-                        public void remove() {
-                            i.remove();
-                        }
-                    };
-                }
+        return new AbstractSet<>() {
+            public Iterator<K> iterator() {
+                return new Iterator<>() {
```

Comment:
There is another PR that proposes to refactor these anonymous classes: https://github.com/openjdk/jdk/pull/15615/files